### PR TITLE
build, deploy: Add support for multi-architecture base images

### PR DESCRIPTION
**Pull request by pdcastro**
Resolves: #1408
Resolves: #1508
Change-type: minor

### CODE REVIEW

File: lib/utils/qemu.ts [hunk 1]
```diff
@@ -84,47 +86,50 @@ export const getQemuPath = function (arch: string) {
...
+				.pipe(extract)
+				.on('error', reject)
+				.on('finish', function () {
+					fs.chmodSync(qemuPath, '755');
+					resolve();
+				});
+		});
+	} catch (err) {
+		try {
+			await fs.promises.unlink(qemuPath);
```
Comment (pdcastro) [inline lib/utils/qemu.ts:127]: The `fs.createWriteStream(qemuPath)` statement towards the beginning of the function creates a zero-length file on disk. On successful download, the file is expanded to the size of the download. On errors, before this PR the file was left behind and then the CLI would assume the file no longer needed to be downloaded (a bug). I've added this `unlink` call to delete "corrupt" files on errors.

File: lib/utils/qemu.ts [hunk 2]
```diff
@@ -148,18 +155,24 @@ export async function installQemuIfNeeded(
...
 		return false;
 	}
 	const { promises: fs } = await import('fs');
 	const qemuPath = await getQemuPath(arch);
 	try {
+		const stats = await fs.stat(qemuPath);
+		// Earlier versions of the CLI with broken error handling would leave
+		// behind files with size 0. If such a file is found, delete it.
+		if (stats.size === 0) {
+			await fs.unlink(qemuPath);
```
Comment (pdcastro) [inline lib/utils/qemu.ts:169]: This is part of a bug fix whereby the CLI would leave behind zero-length files if the user specified an invalid architecture on the command line for the build/deploy commands, e.g. `--arch arm` instead of `--arch armv7hf`. Examples:
```
# ls -l ~/.balena/bin/
total 5476
-rw-r--r-- 1 root root       0 Dec 17 23:57 qemu-execve-386-v4.0.0+balena2
-rwxr-xr-x 1 root root 5606984 Dec 17 23:30 qemu-execve-armv7hf-v4.0.0+balena2
-rw-r--r-- 1 root root       0 Dec 17 23:57 qemu-execve-i386-v4.0.0+balena2
-rw-r--r-- 1 root root       0 Dec 17 23:57 qemu-execve-x86-v4.0.0+balena2
```

File: package.json [hunk 3]
```diff
@@ -256,7 +256,7 @@
     "resin-compose-parse": "^2.1.2",
     "resin-doodles": "^0.1.1",
     "resin-image-fs": "^5.0.9",
-    "resin-multibuild": "^4.7.2",
+    "resin-multibuild": "^4.10.0",
```
Comment (pdcastro) [inline package.json:259]: This one line is the "core feature implementation" really... The rest of the PR is bug fixes (qemu download) and printing hints to the user about the `--pull` and `--emulated` options that become particularly relevant in conjunction with multiarch images.

File: lib/utils/qemu.ts [hunk 4]
```diff
@@ -84,47 +86,50 @@ export const getQemuPath = function (arch: string) {
...
+	const installStream = fs.createWriteStream(qemuPath);
+	try {
+		extract.on('entry', function (header, stream, next) {
+			stream.on('end', next);
+			if (header.name.includes(`qemu-${qemuArch}-static`)) {
+				stream.pipe(installStream);
+			} else {
+				stream.resume();
+			}
+		});
```
Comment (srlowe) [inline lib/utils/qemu.ts:111]: Does having this part under `try` have any effect?  What could throw?

File: lib/utils/qemu.ts [hunk 5]
```diff
@@ -175,30 +188,27 @@ export async function installQemuIfNeeded(
  * - https://stackoverflow.com/questions/55388725/run-linux-arm-container-via-qemu-binfmt-misc-on-docker-lcow
  *
  * @param docker Dockerode instance
+ * @param emulated The --emulated command-line option
+ * @param logger Logger instance
```
Comment (srlowe) [inline lib/utils/qemu.ts:192]: Should we delete the jsdoc params lists?

Comment (pdcastro) [reply lib/utils/qemu.ts:192]: Why, because they are not useful? The `docker` and `logger` descriptions are not useful indeed, but the `emulated` description actually clarifies that it is the `--emulated` command line option.

File: lib/utils/qemu.ts (hunk 4 above)
Comment (pdcastro) [reply lib/utils/qemu.ts:111]: If unnecessary, It is also certainly harmless. I wasn't sure about one specific situation: suppose the execution flow "stopped" at the `await` clause below (`await new Promise(...)`). While "stopped" there, suppose `header.name.includes()` above threw because name is undefined (assume it is possible, though it may not be). In this case, would this `try` block still catch the error? Considering that the error is thrown by a callback of `extract.on`. My gut feeling is that the error would not be caught, but I wasn't sure.

In summary, I think you are right that the `extract.on` block could be moved outside the `try` block, but also harmless to be there. If other changes are needed to the PR I can move this block outside the try-catch, but if this was the only change, I'll just let it be. :-)

File: lib/utils/qemu.ts (hunk 5 above)
Comment (srlowe) [reply lib/utils/qemu.ts:192]: I thought I remember you suggesting we stop adding them, since we are using TS.  I don't mind though.

File: lib/utils/qemu.ts (hunk 4 above)
Comment (srlowe) [reply lib/utils/qemu.ts:111]: > While "stopped" there, suppose header.name.includes() above threw because name is undefined (assume it is possible, though it may not be). In this case, would this try block still catch the error? Considering that the error is thrown by a callback of extract.on. My gut feeling is that the error would not be caught, but I wasn't sure.

Yeah I think that's outside of the context of the try at that point (since the throw is happening in a callback).  I agree it's harmless, but wanted to check that there wasn't a bad assumption.

File: lib/utils/qemu.ts (hunk 5 above)
Comment (pdcastro) [reply lib/utils/qemu.ts:192]: Related to this, what I remember is that some param tags inherited from typed javascript files (.js) include the type hinting that becomes redundant when converted to typescript. Example:

`@param {boolean} emulated The --emulated command-line option`
_vs._
`@param emulated The --emulated command-line option`

So the type hinting should be deleted when converted to Typescript, especially as it introduces the possibility of conflict with the Typescript type specification. But the parameter descriptions may be useful, if they add some value ("Logger instance" does not add value, granted!)

**Comment by ab77** [comment]
@pdcastro @srlowe is fix this blocked by anything? I am also affected by this, via https://github.com/balena-io/balena-cli/issues/2174 cc @klutchell

**Comment by pdcastro** [comment]
> pdcastro srlowe is fix this blocked by anything? I am also affected by this, via #2174 cc klutchell

I've come across complications... Started an FD thread and pinged you guys here:
https://www.flowdock.com/app/rulemotion/i-cli/threads/RuSu1KiWOn62xaGy7O2sn8m8BUc

File: lib/utils/qemu.ts (hunk 1 above)
Comment (pdcastro) [reply lib/utils/qemu.ts]: The commit that this comment referred to was moved to PR #2207.

File: lib/utils/qemu.ts (hunk 2 above)
Comment (pdcastro) [reply lib/utils/qemu.ts]: The commit that this comment referred to was moved to PR #2207.

File: lib/utils/qemu.ts (hunk 4 above)
Comment (pdcastro) [reply lib/utils/qemu.ts]: The commit that this comment referred to was moved to PR #2207.

File: lib/utils/docker.ts [hunk 6]
```diff
@@ -164,12 +164,85 @@ export function generateBuildOpts(options: {
...
-		dockerVersion.Engine && dockerVersion.Engine.match(/balena|balaena/)
+		// dockerVersion.Engine should be 'balena-engine' for the current
+		// version of balenaEngine, but at one point it was spelt 'balaena':
+		// https://github.com/balena-os/balena-engine/pull/32/files
+		(dockerVersion.Engine && dockerVersion.Engine.match(/balena|balaena/))
 	);
 }
+
+/** Detect whether the docker daemon is Docker Desktop (Windows or Mac) */
+export async function isDockerDesktop(
```
Comment (pdcastro) [inline lib/utils/docker.ts:179]: This function was extracted / refactored from `lib/utils/qemu.ts`.

File: lib/utils/compose_ts.ts [hunk 7]
```diff
@@ -907,6 +908,114 @@ export function printGitignoreWarn(
...
+			If you experience that error, please take any one of the following actions:
+			* Upgrade Docker Engine to version 20.10.4 or later. If you are using
+			  Docker Desktop for Mac or Windows, upgrade it to version 3.2.1 or later.
+			* Downgrade Docker Engine to version 19.X.X. If you are using Docker Desktop
+			  for Mac or Windows, downgrade it to version 2.X.X.
+			* Downgrade the balena CLI to v12.40.3 or earlier. This would however cause
+			  support for multi-architecture base images to be lost.
+			* Manually run the 'docker pull' command for all base images listed in your
+			  Dockerfile(s) prior to executing the 'balena build' or 'balena deploy'
+			  commands, and then do not use the balena CLI's '--pull' flag.
```
Comment (pdcastro) [inline lib/utils/compose_ts.ts:977]: Just learned that this advice is insufficient / incorrect with Docker v20.10.4 and v20.10.5. This PR will require further changes. :-/
https://www.flowdock.com/app/rulemotion/i-cli/threads/RuSu1KiWOn62xaGy7O2sn8m8BUc
